feat(appset): add DefaultPrune support to RollingSync (#25444)

fsdkibe opened the pull request:
<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Fixes #25444

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

fsdkibe committed:
chore: update generated code, manifests, and mocks

Signed-off-by: Duncan Kibet <duncankibet.dev@gmail.com>

dependabot[bot] committed:
chore(deps): bump github.com/spf13/cobra from 1.10.1 to 1.10.2 (#25505)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

fsdkibe committed:
Merge branch 'master' into fix/issue-25444-rolling-sync-prune

fsdkibe committed:
Merge branch 'master' into fix/issue-25444-rolling-sync-prune

fsdkibe committed:
Merge branch 'master' into fix/issue-25444-rolling-sync-prune

ranakan19 commented on docs/operator-manual/applicationset/Progressive-Syncs.md:
```diff
@@ -96,6 +96,50 @@ Once each batch of Applications reaches a `Healthy` status, the next batch is sy
 
+#### Resource Pruning with RollingSync
+
+When using RollingSync, resources that are removed from Git are not automatically deleted from the cluster by default. This can cause applications to remain in an `OutOfSync` state indefinitely.
```
>This can cause applications to remain in an `OutOfSync` state indefinitely.

that may not be true, especially **indefinitely**. 
There are workaround pointed in the issue as well (manual sync, let's frame this better so that it does not cause confusion.

ranakan19 reviewed (commented):
Thank you for your contribution!
Were you able to reproduce the issue? I was not - https://github.com/argoproj/argo-cd/issues/25444#issuecomment-3797143740. 
This PR does make it easier to enable pruning at Application level, while still overriding with Application-level setting, unsure though if it helps with issue.

fsdkibe committed:
refactor: update RollingSync resource pruning explanation

Clarified default behavior of RollingSync regarding resource deletion and OutOfSync status.

Signed-off-by: Duncan Kibet <duncankibet.dev@gmail.com>

ranakan19 commented on applicationset/controllers/applicationset_controller.go:
```diff
@@ -1627,8 +1648,11 @@ func syncApplication(application argov1alpha1.Application, prune bool) argov1alp
 	}
+
+	// Set this regardless of whether SyncPolicy exists.
+	operation.Sync.Prune = prune
```
it looks like apart from adding the comment, you also removed it out of the `if application.Spec.SyncPolicy != nil {}` block. I was asking for your reasoning there.

nitishfy commented:
No activity, feel free to reopen this.
